# Scipy 라이브러리
- scipy는 과학 및 수학 문제 해결을 목적으로 사용되는 오픈소스 python 라이브러리
- Numpy 를 기반으로 한 데이터를 조작 및 시각화 패키지
- `scipy.stats`라는 하위 패키지에서 다양한 분포에 대한 통계 함수를 제공
    - 이산형 분포: 이항분포, 포아송분포,...
    - 연속형 분포: 정규분포, t-분포, F-분포, 카이제곱분포,...
- 이 외에도 여러 하위 패키지가 있음(클러스터링, 주파수 등)

### scipy.stats
대표적인  확률분포 이름
- norm: 정규분포
- t: t-분포
- f: F-분포
- chi2: 카이제곱 분포
- bernoulli: 베르누이 분포
- binom: 이항분포
- poisson: 포아송 분포
- uniform: 균일분포
- expon: 지수분포

확률분포 클래스의 메서드
- pdf: 확률밀도함수 (probability density function)
- cdf: 누적분포함수 (cumulative distribution function)
- ppf: 누적분포함수의 역함수 (inverse cumulative distribution function)
- rvs: 랜덤 표본 생성 (random variates sampling) 

*메서드: 객체지향 프로그래밍(자주 쓰는 특별한 구조들을 클래스라는 것으로 정의하고 클래스 안에 여러 변수와 함수를 사용하게 하는 것)에서 클래스 안에 포함된 함수   
*메서드나 클래스의 변수를 사용할 때에는 새로운 객체(오브젝트)를 만들어서 클래스 지정 등으로 활용하게 됨. 객체에 메서드를 붙여서 활용

### 정규분포

확률 밀도 함수를 볼 때 미지수(모르는 값, 모수의 값)는 시그마, 뮤 값임(평균과 분산)
확률 변수 x의 값은 알려진 값이다.   
확률 변수도 특정 범위를 가지는 경우가 있다. 카이제곱이나 F-분포 같은 경우는 음수값을 가질 수 없고 0부터 무한대의 값을 확률 변수가 가질 수 있다.   
시그마도 표준편차이기 때문에 항상 양수여야 함.   
표준 정규분포: 평균이 0이고 분산이 1인 정규분포. z로 보통 표기함 

### 확률밀도함수 계산
- scipy.stats.norm.pdf(x,loc=0,scale=1)
    - x: 확률변수의 값
    - loc: 평균(뮤)
    - scale: 표준편차(시그마)
    
*Python 개념 계층 구조 
Library  
└── Package  
    └── Module (.py file)  
        ├── Class  
        │   ├── Method (function)  
        │   └── Attribute (variable)  
        ├── Function  
        └── Variable

*scipy는 라이브러리. scipy 내부에 stats라는 패키지가 존재하는 것

In [12]:
#scipy 라이브러리 안에 있는 stats 패키지를 import
from scipy import stats
#stats 패키지 안에 있는 norm 클래스(함수)를 정의
#norm클래스 안에 있는 pdf라는 메서드를 불러와 사용
#확률 변수의 값이 0일 때, N(0,1)의 확률밀도함수 값 계산
stats.norm.pdf(0,loc=0, scale=1)
#loc, scale은 매개 변수로 라이브러리마다 이름이 다를 수 있음

np.float64(0.3989422804014327)

scipy는 numpy를 기본 베이스로 가져가서 결과값에 np가 나타남   

In [13]:
#확률 변수 값을 리스트로 만들어서 pdf 메서드에 넣으면 각 확률밀도함수 값을 계산해줌
stats.norm.pdf([0,1,2],loc=0, scale=1)

array([0.39894228, 0.24197072, 0.05399097])

In [14]:
#누적분포함수
stats.norm.cdf(10,loc=10, scale=5)

np.float64(0.5)

| 함수                  | 의미             | 예시 결과      |
| ------------------- | -------------- | ---------- |
| `stats.norm.pdf(0)` | x=0에서의 밀도 (높이) | `≈ 0.3989` |
| `stats.norm.cdf(0)` | x≤0일 확률 (누적)   | `0.5`      |


누적분포함수의 역함수
- 특정 확률값에 해당하는 확률변수의 값을 계산
- 추후 가설검정에서 유의수준이 a(알파)일 때 확률변수의 값(임계치)를 계산하는 데 사용 

In [15]:
stats.norm.ppf(0.5,loc=10,scale=5)

np.float64(10.0)

In [16]:
stats.norm.ppf([0.05,0.5,0.95],loc=10,scale=5)

array([ 1.77573187, 10.        , 18.22426813])

### Numpy 라이브러리

np.set_printoptions(precision=n): 소수점 4자리까지 표기   
np.set_printoptions(supress=True): 지수표기법 사용하지 않고 일반 소수점 표기

In [17]:
import matplotlib.pyplot as plt
import numpy as np

import koreanize_matplotlib #한글 깨짐 방지


Numpy 라이브러리 사용하면 반복문 없이 선형대수 관련된 빠른 연산 가능   
np.sum() : 합계   
np.cumprod : 누적곱     
np.cumsum : 누적합  
np.prod : 곱    
np.argmin, np.argmax : 최소값, 최대값의 인덱스  
np.diff: 차분, 빼는 것  

In [ ]:
import numpy as np
np.set_printoptions(suppress=True, precision=2)

arr=[1,10,20,15,5]
print(f'{np.sum(arr) = }')
print(f'{np.prod(arr) = }')
print(f'{np.cumsum(arr)=}')
print(f'{np.cumprod(arr)=}')
print(f'{np.diff(arr)=}')
print(f'{np.exp(arr)=}') #set_printoptions(suppress=True, precision=2)로 설정했는데 적용 안되게 출력됨.. 

np.sum(arr) = np.int64(51)
np.prod(arr) = np.int64(15000)
np.cumsum(arr)=array([ 1, 11, 31, 46, 51])
np.cumprod(arr)=array([    1,    10,   200,  3000, 15000])
np.diff(arr)=array([  9,  10,  -5, -10])
np.exp(arr)=array([2.72e+00, 2.20e+04, 4.85e+08, 3.27e+06, 1.48e+02])


기술통계량 계산 함수    
- np.mean()   
- np.median() 
- np.var()    
    - ddof=1 : 비편향 분산 계산, 표본분산 구할 때 항상 지정 해줘야 함(모분산을 구할 때는 지정 안 해줘도 됨, 디폴드가 0임)   
- IQR 구하는 법
    - q1,q3=np.percentile(data,[25,75])   #백분위수 계산    
    - q1,q3=np.quantile(data,[0.25,0.75])   #사분위수 계산   
    - IQR = q3-q1

### Pandas 라이브러리
데이터 프레임 형식(Dataframe)과 시리즈(Series)를 다룰 수 있게 해주는 라이브러리

pd.options.display.float_format ='{:.2f}'.format #데이터값 실수, 소수점 둘째자리까지 표시 설정 

In [ ]:
import pandas as pd
pd.options.display.float_format = '{:.2f}'.format
df=pd.DataFrame({
    'name': ['김홍익','이영희','박철수','최영수','이순신','강감찬', '홍길동','신사임당','유관순','윤봉길'],
    'height': [170, 160, 180, 175, 165, 185, 172, 168, 178, 177],
    'weight': [70, 60, 80, 75, 65, 85, 72, 68, 78, 77]
})
display(df) #print(df)를 하면 문자열로 출력됨

name  height  weight
0   김홍익     170      70
1   이영희     160      60
2   박철수     180      80
3   최영수     175      75
4   이순신     165      65
5   강감찬     185      85
6   홍길동     172      72
7  신사임당     168      68
8   유관순     178      78
9   윤봉길     177      77

In [ ]:
df.mean(numeric_only=True) 
#numeric_only=True라고 숫자형 데이터만 계산하라고 지시해야 함

height   173.00
weight    73.00
dtype: float64

In [ ]:
display(df.info())
#object 타입은 문자열
#int64는 정수형, float64는 실수형
#df의 type은 pandas.core.frame.DataFrame으로 하나의 객체로 정의됨
#객체는 여러 속성과 메서드를 가짐, 어떤 메서드를 사용할 수 있는지 확인하려면 dir(df)를 사용하면 됨
display(dir(df))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   name    10 non-null     object
 1   height  10 non-null     int64 
 2   weight  10 non-null     int64 
dtypes: int64(2), object(1)
memory usage: 368.0+ bytes


None

['T',
 '_AXIS_LEN',
 '_AXIS_ORDERS',
 '_AXIS_TO_AXIS_NUMBER',
 '_HANDLED_TYPES',
 '__abs__',
 '__add__',
 '__and__',
 '__annotations__',
 '__array__',
 '__array_priority__',
 '__array_ufunc__',
 '__arrow_c_stream__',
 '__bool__',
 '__class__',
 '__contains__',
 '__copy__',
 '__dataframe__',
 '__dataframe_consortium_standard__',
 '__deepcopy__',
 '__delattr__',
 '__delitem__',
 '__dict__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__finalize__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__iand__',
 '__ifloordiv__',
 '__imod__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__invert__',
 '__ior__',
 '__ipow__',
 '__isub__',
 '__iter__',
 '__itruediv__',
 '__ixor__',
 '__le__',
 '__len__',
 '__lt__',
 '__matmul__',
 '__mod__',
 '__module__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__nonzero__',
 '__or__',
 '__pandas_priority__',
 '__pos__',
 '__pow__',
 '__r

In [38]:
df[['weight', 'height']].median()  #여러 열을 선택 가능 

weight    73.50
height   173.50
dtype: float64

기술 통계량 계산
- Numpy에서는 ddof=0 디폴트였지만 Pandas에서는 ddof=1(자유도)가 디폴트임
- 모표준편차를 구할 때 ddof=0 지정해줘야 함 

In [40]:
print(df.std(numeric_only=True))  #표준편차 계산
print(df.std(numeric_only=True, ddof=0))  #모집단 표준편차 계산, ddof=0은 모집단 표준편차를 의미함
#결과값은 표본표준편차가 n-1로 계산하기 때문에 모표준편차보다 약간 크게 나옴 

height   7.50
weight   7.50
dtype: float64
height   7.11
weight   7.11
dtype: float64


pandas에서 상관계수와 공분산도 계산해줌
- df.corr(): 피어슨상관계수
- df.cov() : 공분산(상관계수를 0과 1 사이의 값으로 표준화한 것)

데이터프레임의 요약 기술통계량 계산
- df.describe(): 데이터프레임의 요약 통계량 계산 

In [ ]:
df.describe() #기본 통계량을 계산해줌
#describe() 메서드는 기본적으로 숫자형 데이터에 대해서만 계산함

height  weight
count   10.00   10.00
mean   173.00   73.00
std      7.50    7.50
min    160.00   60.00
25%    168.50   68.50
50%    173.50   73.50
75%    177.75   77.75
max    185.00   85.00

### Seaborn 라이브러리
matplotlib기반의 python 데이터 시각화 라이브러리    
seaborn 사이트의 gallery에 들어가면 그릴 수 있는 그래프들이 나옴

In [42]:
import seaborn as sns